File: exporter-app/utils/scripts/generate-link.js
#!/usr/bin/env node

/**
 * Helper script to generate a link to the exporter app with a pre-loaded diagram
 * 
 * Usage:
 *   node generate-link.js "sequenceDiagram\nA->>B: Hello"
 *   node generate-link.js < diagram.mmd
 * 
 * Or pipe Mermaid code:
 *   echo "sequenceDiagram\nA->>B: Hello" | node generate-link.js
 */

const readline = require('readline');

function generateLink(mermaidCode) {
  if (!mermaidCode || !mermaidCode.trim()) {
    console.error('Error: No Mermaid code provided');
    process.exit(1);
  }

  // Base64 encode the diagram code
  const encoded = Buffer.from(mermaidCode.trim()).toString('base64');
  
  // URL-encode the Base64 string to handle special characters (+, /, =)
  const urlEncoded = encodeURIComponent(encoded);
  
  // Default to localhost:8080, but can be overridden with PORT env var
  const port = process.env.PORT || 8080;
  const host = process.env.HOST || 'localhost';
  
  const url = `http://${host}:${port}/?code=${urlEncoded}`;
  
  return url;
}

// Check if code is provided as command line argument
if (process.argv.length > 2) {
  const mermaidCode = process.argv.slice(2).join('\n');
  const link = generateLink(mermaidCode);
  console.log(link);
  process.exit(0);
}

// Otherwise, read from stdin
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

let input = '';

rl.on('line', (line) => {
  input += line + '\n';
});

rl.on('close', () => {
  const link = generateLink(input);
  console.log(link);
  process.exit(0);
});

File: exporter-app/utils/scripts/generate_link.py
#!/usr/bin/env python3
"""
Helper script to generate a link to the exporter app with a pre-loaded diagram

Usage:
    python generate_link.py "sequenceDiagram\nA->>B: Hello"
    python generate_link.py < diagram.mmd
    cat diagram.mmd | python generate_link.py
"""

import sys
import base64
import os

def generate_link(mermaid_code, host='localhost', port=8080):
    """
    Generate a URL link to the exporter app with pre-loaded diagram
    
    Args:
        mermaid_code: The Mermaid diagram code as a string
        host: The hostname (default: localhost)
        port: The port number (default: 8080)
    
    Returns:
        The full URL with encoded diagram
    """
    if not mermaid_code or not mermaid_code.strip():
        raise ValueError("No Mermaid code provided")
    
    # Base64 encode the diagram code
    encoded = base64.b64encode(mermaid_code.strip().encode('utf-8')).decode('utf-8')
    
    # Construct the URL
    url = f"http://{host}:{port}/?code={encoded}"
    
    return url

if __name__ == "__main__":
    # Get host and port from environment variables if available
    host = os.getenv('HOST', 'localhost')
    port = int(os.getenv('PORT', 8080))
    
    # Check if code is provided as command line argument
    if len(sys.argv) > 1:
        mermaid_code = '\n'.join(sys.argv[1:])
        link = generate_link(mermaid_code, host, port)
        print(link)
    else:
        # Read from stdin
        mermaid_code = sys.stdin.read()
        link = generate_link(mermaid_code, host, port)
        print(link)

